From e6a31225d6c4aa50ada42efba93e53a394b2a8b0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Fri, 13 Sep 2024 12:54:12 +0200
Subject: [PATCH 1148/1695] drivers: stm32_rtc: add alarm related operations
 and init function
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The RTC framework allows the registration of function :
- read_alarm
- set_alarm
- alarm_enable
- wait_alarm
- cancel_wait

The purpose of the init function is to configure and setup the RTC
at probe time: it configures the prescalers, and make sure the time is
in 24h format.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 05b7a4a77c31b077283cb43900b52b41ef7d6ee3 drivers: stm32_rtc: add alarm related operations and init function]
Change-Id: Id2eed1711616b25f49e48228a40d0d65ade97b0c
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/402362
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_rtc.c | 408 ++++++++++++++++++++++++++++++++++++++-
 1 file changed, 407 insertions(+), 1 deletion(-)

diff --git a/core/drivers/stm32_rtc.c b/core/drivers/stm32_rtc.c
index 6e772aa96..19f3b1b57 100644
--- a/core/drivers/stm32_rtc.c
+++ b/core/drivers/stm32_rtc.c
@@ -8,13 +8,17 @@
 #else
 #include <arm32.h>
 #endif /* CFG_ARM64_core */
+#include <config.h>
 #include <drivers/clk.h>
 #include <drivers/clk_dt.h>
 #include <drivers/rtc.h>
 #include <drivers/stm32_rtc.h>
 #include <drivers/stm32_rif.h>
 #include <io.h>
+#include <keep.h>
 #include <kernel/dt.h>
+#include <kernel/interrupt.h>
+#include <kernel/notif.h>
 #include <kernel/panic.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
@@ -82,10 +86,16 @@
 #define RTC_ICSR_INITF			BIT(6)
 #define RTC_ICSR_INIT			BIT(7)
 
+#define RTC_PRER_PREDIV_S_SHIFT		U(0)
 #define RTC_PRER_PREDIV_S_MASK		GENMASK_32(14, 0)
+#define RTC_PRER_PREDIV_A_SHIFT		U(16)
+#define RTC_PRER_PREDIV_A_MASK		GENMASK_32(22, 16)
 
 #define RTC_CR_BYPSHAD			BIT(5)
 #define RTC_CR_BYPSHAD_SHIFT		U(5)
+#define RTC_CR_FMT			BIT(6)
+#define RTC_CR_ALRAE			BIT(8)
+#define RTC_CR_ALRAIE			BIT(12)
 #define RTC_CR_TAMPTS			BIT(25)
 
 #define RTC_PRIVCFGR_VALUES		GENMASK_32(3, 0)
@@ -114,9 +124,32 @@
 #define RTC_TSDR_DU_MASK		GENMASK_32(3, 0)
 #define RTC_TSDR_DU_SHIFT		U(0)
 
+#define RTC_ALRMXR_SEC_UNITS_MASK	GENMASK_32(3, 0)
+#define RTC_ALRMXR_SEC_UNITS_SHIFT	U(0)
+#define RTC_ALRMXR_SEC_TENS_MASK	GENMASK_32(6, 4)
+#define RTC_ALRMXR_SEC_TENS_SHIFT	U(4)
+#define RTC_ALRMXR_SEC_MASK		BIT(7)
+#define RTC_ALRMXR_MIN_UNITS_MASK	GENMASK_32(11, 8)
+#define RTC_ALRMXR_MIN_UNITS_SHIFT	U(8)
+#define RTC_ALRMXR_MIN_TENS_MASK	GENMASK_32(14, 12)
+#define RTC_ALRMXR_MIN_TENS_SHIFT	U(12)
+#define RTC_ALRMXR_MIN_MASK		BIT(15)
+#define RTC_ALRMXR_HOUR_UNITS_MASK	GENMASK_32(19, 16)
+#define RTC_ALRMXR_HOUR_UNITS_SHIFT	U(16)
+#define RTC_ALRMXR_HOUR_TENS_MASK	GENMASK_32(21, 20)
+#define RTC_ALRMXR_HOUR_TENS_SHIFT	U(20)
+#define RTC_ALRMXR_PM			BIT(22)
+#define RTC_ALRMXR_HOUR_MASK		BIT(23)
+#define RTC_ALRMXR_DATE_UNITS_MASK	GENMASK_32(27, 24)
+#define RTC_ALRMXR_DATE_UNITS_SHIFT	U(24)
+#define RTC_ALRMXR_DATE_TENS_MASK	GENMASK_32(29, 28)
+#define RTC_ALRMXR_DATE_TENS_SHIFT	U(28)
+
+#define RTC_SR_ALRAF			BIT(0)
 #define RTC_SR_TSF			BIT(3)
 #define RTC_SR_TSOVF			BIT(4)
 
+#define RTC_SCR_CALRAF			BIT(0)
 #define RTC_SCR_CTSF			BIT(3)
 #define RTC_SCR_CTSOVF			BIT(4)
 
@@ -154,6 +187,22 @@ struct rtc_compat {
 	bool has_rif_support;
 };
 
+/*
+ * struct rtc_device - RTC device structure
+ * @base: Base address of the RTC device
+ * @compat: RTC device compatibility
+ * @pclk: RTC peripheral clock
+ * @rtc_ck: RTC core clock
+ * @conf_data: array of RIF configuration data
+ * @nb_res: Number of RIF configurations
+ * @flags: RTC device flags
+ * @is_secured: RTC device secured status
+ * @itr_chip: Interrupt chip
+ * @itr_num: Interrupt number
+ * @itr_handler: Interrupt handler
+ * @notif_id: Notification ID
+ * @wait_alarm_return_status: Status of the wait alarm
+ */
 struct rtc_device {
 	struct io_pa_va base;
 	struct rtc_compat compat;
@@ -163,6 +212,11 @@ struct rtc_device {
 	unsigned int nb_res;
 	uint8_t flags;
 	bool is_secured;
+	struct itr_chip *itr_chip;
+	size_t itr_num;
+	struct itr_handler *itr_handler;
+	uint32_t notif_id;
+	enum rtc_wait_alarm_status wait_alarm_return_status;
 };
 
 /* Expect a single RTC instance */
@@ -667,6 +721,32 @@ static void apply_rif_config(bool is_tdcid)
 	}
 }
 
+static void stm32_rtc_clear_events(uint32_t flags)
+{
+	io_write32(get_base() + RTC_SCR, flags);
+}
+
+static enum itr_return stm32_rtc_it_handler(struct itr_handler *h __unused)
+{
+	vaddr_t rtc_base = get_base();
+	uint32_t status = io_read32(rtc_base + RTC_SR);
+	uint32_t cr = io_read32(rtc_base + RTC_CR);
+
+	if ((status & RTC_SR_ALRAF) && (cr & RTC_CR_ALRAIE)) {
+		DMSG("Alarm occurred");
+		/* Clear event's flags */
+		stm32_rtc_clear_events(RTC_SCR_CALRAF);
+		/*
+		 * Notify the caller of 'stm32_rtc_wait_alarm' to re-schedule
+		 * the calling thread
+		 */
+		notif_send_async(rtc_dev.notif_id);
+	}
+
+	return ITRR_HANDLED;
+}
+DECLARE_KEEP_PAGER(stm32_rtc_it_handler);
+
 static TEE_Result parse_dt(const void *fdt, int node)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
@@ -691,6 +771,20 @@ static TEE_Result parse_dt(const void *fdt, int node)
 	if (!rtc_dev.rtc_ck)
 		return res;
 
+	if (IS_ENABLED(CFG_RTC_PTA) && IS_ENABLED(CFG_DRIVERS_RTC)) {
+		res = interrupt_dt_get(fdt, node, &rtc_dev.itr_chip,
+				       &rtc_dev.itr_num);
+		if (res)
+			return res;
+		res = interrupt_create_handler(rtc_dev.itr_chip,
+					       rtc_dev.itr_num,
+					       stm32_rtc_it_handler,
+					       &rtc_dev, 0,
+					       &rtc_dev.itr_handler);
+		if (res)
+			return res;
+	}
+
 	if (!rtc_dev.compat.has_rif_support)
 		return res;
 
@@ -768,6 +862,81 @@ static TEE_Result stm32_rtc_wait_sync(void)
 	return TEE_SUCCESS;
 }
 
+static TEE_Result stm32_rtc_init(void)
+{
+	unsigned long rate = clk_get_rate(rtc_dev.rtc_ck);
+	vaddr_t base = get_base();
+	uint32_t pred_a = 0;
+	uint32_t pred_s = 0;
+	uint32_t prer = io_read32(base + RTC_PRER);
+	uint32_t cr = io_read32(base + RTC_CR);
+	uint32_t pred_a_max = RTC_PRER_PREDIV_A_MASK >> RTC_PRER_PREDIV_A_SHIFT;
+	uint32_t pred_s_max = RTC_PRER_PREDIV_S_MASK >> RTC_PRER_PREDIV_S_SHIFT;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (rate > (pred_a_max + 1) * (pred_s_max + 1))
+		panic("rtc_ck rate is too high");
+
+	/*
+	 * Compute the prescaler values whom divides the clock in order to get a
+	 * 1 Hz output
+	 */
+	for (pred_a = 0; pred_a <= pred_a_max; pred_a++) {
+		pred_s = (rate / (pred_a + 1)) - 1;
+		if (pred_s <= pred_s_max &&
+		    ((pred_s + 1) * (pred_a + 1)) == rate)
+			break;
+	}
+	/*
+	 * Can't find a 1Hz, so give priority to RTC power consumption
+	 * by choosing the higher possible value for prediv_a
+	 */
+	if (pred_s > pred_s_max || pred_a > pred_a_max) {
+		pred_a = pred_a_max;
+		pred_s = (rate / (pred_a + 1)) - 1;
+
+		DMSG("rtc_ck is %s",
+		     (rate < ((pred_a + 1) * (pred_s + 1))) ? "fast" : "slow");
+	}
+
+	prer &= RTC_PRER_PREDIV_S_MASK | RTC_PRER_PREDIV_A_MASK;
+	pred_s = (pred_s << RTC_PRER_PREDIV_S_SHIFT) & RTC_PRER_PREDIV_S_MASK;
+	pred_a = (pred_a << RTC_PRER_PREDIV_A_SHIFT) & RTC_PRER_PREDIV_A_MASK;
+	/* quit if there is nothing to initialize */
+	if ((cr & RTC_CR_FMT) == 0 && prer == (pred_s | pred_a))
+		return TEE_SUCCESS;
+
+	stm32_rtc_write_unprotect();
+
+	res = stm32_rtc_enter_init_mode();
+	if (res) {
+		EMSG("Can't enter in init mode. Prescaler config failed");
+		stm32_rtc_write_protect();
+		return TEE_ERROR_BUSY;
+	}
+
+	io_write32(base + RTC_PRER, pred_s);
+	io_write32(base + RTC_PRER, pred_a | pred_s);
+
+	/* Force 24h time format */
+	cr &= ~RTC_CR_FMT;
+
+	io_write32(base + RTC_CR, cr);
+
+	stm32_rtc_exit_init_mode();
+
+	res = stm32_rtc_wait_sync();
+	if (res) {
+		EMSG("Can't sync RTC. Prescaler config failed");
+		stm32_rtc_write_protect();
+		return TEE_ERROR_BUSY;
+	}
+
+	stm32_rtc_write_protect();
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result stm32_rtc_get_time(struct rtc *rtc __unused,
 				     struct optee_rtc_time *tm)
 {
@@ -847,9 +1016,230 @@ static TEE_Result stm32_rtc_set_time(struct rtc *rtc, struct optee_rtc_time *tm)
 	return TEE_SUCCESS;
 }
 
+static TEE_Result stm32_rtc_read_alarm(struct rtc *rtc,
+				       struct optee_rtc_alarm *alarm)
+{
+	struct optee_rtc_time *alarm_tm = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct optee_rtc_time current_tm = { };
+	vaddr_t rtc_base = get_base();
+	uint32_t alrmar = io_read32(rtc_base + RTC_ALRMAR);
+	uint32_t cr = io_read32(rtc_base + RTC_CR);
+	uint32_t status = io_read32(rtc_base + RTC_SR);
+
+	alarm_tm = &alarm->time;
+
+	res = stm32_rtc_get_time(rtc, &current_tm);
+	if (res)
+		return res;
+
+	alarm_tm->tm_year = current_tm.tm_year;
+	alarm_tm->tm_mon = current_tm.tm_mon;
+	alarm_tm->tm_mday = ((alrmar & RTC_ALRMXR_DATE_UNITS_MASK) >>
+			    RTC_ALRMXR_DATE_UNITS_SHIFT) +
+			    ((alrmar & RTC_ALRMXR_DATE_TENS_MASK) >>
+			    RTC_ALRMXR_DATE_TENS_SHIFT) * 10;
+	alarm_tm->tm_hour = ((alrmar & RTC_ALRMXR_HOUR_UNITS_MASK) >>
+			    RTC_ALRMXR_HOUR_UNITS_SHIFT) +
+			    ((alrmar & RTC_ALRMXR_HOUR_TENS_MASK) >>
+			    RTC_ALRMXR_HOUR_TENS_SHIFT) * 10;
+	alarm_tm->tm_min = ((alrmar & RTC_ALRMXR_MIN_UNITS_MASK) >>
+			    RTC_ALRMXR_MIN_UNITS_SHIFT) +
+			   ((alrmar & RTC_ALRMXR_MIN_TENS_MASK) >>
+			    RTC_ALRMXR_MIN_TENS_SHIFT) * 10;
+	alarm_tm->tm_sec = ((alrmar & RTC_ALRMXR_MIN_UNITS_MASK) >>
+			    RTC_ALRMXR_MIN_UNITS_SHIFT) +
+			   ((alrmar & RTC_ALRMXR_MIN_TENS_MASK) >>
+			    RTC_ALRMXR_MIN_TENS_SHIFT) * 10;
+
+	if (rtc_timecmp(alarm_tm, &current_tm) < 0) {
+		if (current_tm.tm_mon == 11) {
+			alarm_tm->tm_mon = 0;
+			alarm_tm->tm_year += 1;
+		} else {
+			alarm_tm->tm_mon += 1;
+		}
+	}
+
+	alarm->enabled = cr & RTC_CR_ALRAE;
+	alarm->pending = status & RTC_SR_ALRAF;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rtc_enable_alarm(struct rtc *rtc __unused, bool enabled)
+{
+	vaddr_t rtc_base = get_base();
+
+	stm32_rtc_write_unprotect();
+
+	if (enabled)
+		io_setbits32(rtc_base + RTC_CR, RTC_CR_ALRAIE | RTC_CR_ALRAE);
+	else
+		io_clrbits32(rtc_base + RTC_CR, RTC_CR_ALRAIE | RTC_CR_ALRAE);
+
+	stm32_rtc_clear_events(RTC_SCR_CALRAF);
+
+	stm32_rtc_write_protect();
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rtc_valid_alarm_time(struct optee_rtc_time *tm)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct stm32_rtc_calendar cal = { };
+	struct optee_rtc_time max = { };
+	uint32_t next_month = 0;
+	uint32_t next_year = 0;
+
+	/*
+	 * Assuming current date is M-D-Y H:M:S.
+	 * RTC alarm can't be set on a specific month and year.
+	 * So the valid alarm range is:
+	 *	M-D-Y H:M:S < alarm <= (M+1)-D-Y H:M:S
+	 */
+	res = stm32_rtc_get_calendar(&cal);
+	if (res)
+		return res;
+
+	stm32_rtc_fill_time(&cal, &max);
+	stm32_rtc_fill_date(&cal, &max);
+	/*
+	 * Hardware month is from 1 to 12
+	 * align with tm_mon for comparison with alarm time
+	 */
+	max.tm_mon -= 1;
+
+	/*
+	 * Don't allow alarm to be set in the past.
+	 */
+	if (rtc_timecmp(&max, tm) > 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/*
+	 * Find the next month and the year of the next month.
+	 * Note: tm_mon and next_month are from 0 to 11
+	 */
+	next_month = max.tm_mon + 1;
+	if (next_month == 12) {
+		next_month = 0;
+		next_year = max.tm_year + 1;
+	} else {
+		next_year = max.tm_year;
+	}
+
+	if (max.tm_mday < tm->tm_mday) {
+		/*
+		 * Alarm is next month
+		 * Must be sure it does not exceed the number of days in
+		 * the next month.
+		 */
+		if (tm->tm_mday > rtc_get_month_days(next_month, next_year))
+			return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	max.tm_year = next_year;
+	max.tm_mon = next_month;
+
+	if (rtc_timecmp(&max, tm) < 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rtc_set_alarm(struct rtc *rtc,
+				      struct optee_rtc_alarm *alarm)
+{
+	struct optee_rtc_time *alarm_time = NULL;
+	vaddr_t rtc_base = get_base();
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	uint32_t alrmar = 0;
+	uint32_t cr = io_read32(rtc_base + RTC_CR);
+
+	alarm_time = &alarm->time;
+
+	res = stm32_rtc_valid_alarm_time(alarm_time);
+	if (res)
+		return res;
+
+	alrmar = 0;
+	/* tm_year and tm_mon are not used because not supported by RTC */
+	alrmar |= ((alarm_time->tm_mday / 10) << RTC_ALRMXR_DATE_TENS_SHIFT) &
+		  RTC_ALRMXR_DATE_TENS_MASK;
+	alrmar |= ((alarm_time->tm_mday % 10) << RTC_ALRMXR_DATE_UNITS_SHIFT) &
+		  RTC_ALRMXR_DATE_UNITS_MASK;
+	/* 24-hour format */
+	alrmar &= ~RTC_ALRMXR_PM;
+	alrmar |= ((alarm_time->tm_hour / 10) << RTC_ALRMXR_HOUR_TENS_SHIFT) &
+		  RTC_ALRMXR_HOUR_TENS_MASK;
+	alrmar |= ((alarm_time->tm_hour % 10) << RTC_ALRMXR_HOUR_UNITS_SHIFT) &
+		  RTC_ALRMXR_HOUR_UNITS_MASK;
+	alrmar |= ((alarm_time->tm_min / 10) << RTC_ALRMXR_MIN_TENS_SHIFT) &
+		  RTC_ALRMXR_MIN_TENS_MASK;
+	alrmar |= ((alarm_time->tm_min % 10) << RTC_ALRMXR_MIN_UNITS_SHIFT) &
+		  RTC_ALRMXR_MIN_UNITS_MASK;
+	alrmar |= ((alarm_time->tm_sec / 10) << RTC_ALRMXR_SEC_TENS_SHIFT) &
+		  RTC_ALRMXR_SEC_TENS_MASK;
+	alrmar |= ((alarm_time->tm_sec % 10) << RTC_ALRMXR_SEC_UNITS_SHIFT) &
+		  RTC_ALRMXR_SEC_UNITS_MASK;
+
+	stm32_rtc_write_unprotect();
+
+	/* Disable Alarm */
+	cr &= ~RTC_CR_ALRAE;
+	io_write32(rtc_base + RTC_CR, cr);
+
+	io_write32(rtc_base + RTC_ALRMAR, alrmar);
+
+	stm32_rtc_enable_alarm(rtc, alarm->enabled);
+
+	stm32_rtc_write_protect();
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rtc_cancel_wait_alarm(struct rtc *rtc __unused)
+{
+	rtc_dev.wait_alarm_return_status = RTC_WAIT_ALARM_CANCELED;
+	notif_send_async(rtc_dev.notif_id);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result
+stm32_rtc_wait_alarm(struct rtc *rtc __unused,
+		     enum rtc_wait_alarm_status *return_status)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	rtc_dev.wait_alarm_return_status = RTC_WAIT_ALARM_RESET;
+
+	/* Wait until a notification arrives - blocking */
+
+	res = notif_wait(rtc_dev.notif_id);
+	if (res)
+		return res;
+
+	if (rtc_dev.wait_alarm_return_status ==
+		RTC_WAIT_ALARM_CANCELED) {
+		*return_status = RTC_WAIT_ALARM_CANCELED;
+		stm32_rtc_enable_alarm(rtc, 0);
+	} else {
+		*return_status = RTC_WAIT_ALARM_ALARM_OCCURRED;
+	}
+	return TEE_SUCCESS;
+}
+
 static const struct rtc_ops stm32_rtc_ops = {
 	.get_time = stm32_rtc_get_time,
 	.set_time = stm32_rtc_set_time,
+	.read_alarm = stm32_rtc_read_alarm,
+	.set_alarm = stm32_rtc_set_alarm,
+	.enable_alarm = stm32_rtc_enable_alarm,
+	.wait_alarm = stm32_rtc_wait_alarm,
+	.cancel_wait = stm32_rtc_cancel_wait_alarm,
 };
 
 static struct rtc stm32_rtc = {
@@ -904,8 +1294,24 @@ static TEE_Result stm32_rtc_probe(const void *fdt, int node,
 		clk_disable(rtc_dev.pclk);
 	}
 
-	if (rtc_dev.is_secured)
+	if (IS_ENABLED(CFG_RTC_PTA) && IS_ENABLED(CFG_DRIVERS_RTC) &&
+	    IS_ENABLED(CFG_CORE_ASYNC_NOTIF) && rtc_dev.is_secured) {
+		res = notif_alloc_async_value(&rtc_dev.notif_id);
+		if (res)
+			return res;
+
+		/* Unbalanced clock enable to ensure IRQ interface is alive */
+		res = clk_enable(rtc_dev.pclk);
+		if (res)
+			return res;
+
+		res = stm32_rtc_init();
+		if (res)
+			return res;
+
 		rtc_register(&stm32_rtc);
+		interrupt_enable(rtc_dev.itr_chip, rtc_dev.itr_num);
+	}
 
 	return res;
 }
-- 
2.47.2

